common/partition/default-partitioner.go

```diff
@@ -0,0 +1,183 @@
+	}
+	return r.IsDrained(ctx, domain.GetInfo().Name, isolationGroup)
+}
+
+func (r *defaultPartitioner) GetIsolationGroupByDomainID(ctx context.Context, domainID string, wfPartitionData PartitionConfig) (*string, error) {
+	if !r.config.zonalPartitioningEnabledGlobally(domainID) {
+		return nil, nil
+	}
+
+	wfPartition := mapPartitionConfigToDefaultPartitionConfig(wfPartitionData)
+	state, err := r.isolationGroupState.GetByDomainID(ctx, domainID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get isolation group state: %w", err)
+	}
+	available := availableIG(r.config.allIsolationGroups, state.Global, state.Domain)
+
+	ig := pickIsolationGroup(wfPartition, available)
+	return &ig, nil
+}
+
+func isDrained(isolationGroup string, global types.IsolationGroupConfiguration, domain types.IsolationGroupConfiguration) bool {
+	globalCfg, hasGlobalConfig := global[string(isolationGroup)]
+	domainCfg, hasDomainConfig := domain[string(isolationGroup)]
+	if hasGlobalConfig {
+		if globalCfg.State == types.IsolationGroupStateDrained {
+			return true
+		}
+	}
+	if hasDomainConfig {
+		if domainCfg.State == types.IsolationGroupStateDrained {
+			return true
+		}
+	}
+	return false
+}
+
+// A simple explicit deny-based isolation group implementation
+func availableIG(allIsolationGroups []string, global types.IsolationGroupConfiguration, domain types.IsolationGroupConfiguration) types.IsolationGroupConfiguration {
+	out := types.IsolationGroupConfiguration{}
+	for _, isolationGroup := range allIsolationGroups {
+		globalCfg, hasGlobalConfig := global[string(isolationGroup)]
+		domainCfg, hasDomainConfig := domain[string(isolationGroup)]
+		if hasGlobalConfig {
+			if globalCfg.State == types.IsolationGroupStateDrained {
+				continue
+			}
+		}
+		if hasDomainConfig {
+			if domainCfg.State == types.IsolationGroupStateDrained {
+				continue
+			}
+		}
+		out[isolationGroup] = types.IsolationGroupPartition{
+			Name:  isolationGroup,
+			State: types.IsolationGroupStateHealthy,
+		}
+	}
+	return out
+}
+
+func mapPartitionConfigToDefaultPartitionConfig(config PartitionConfig) defaultWorkflowPartitionConfig {
+	isolationGroup, _ := config[defaultPartitionConfigWorkerIsolationGroup]
+	runID, _ := config[defaultPartitionConfigRunID]
+	return defaultWorkflowPartitionConfig{
+		WorkflowStartIsolationGroup: isolationGroup,
+		RunID:                       runID,
+	}
+}
+
+// picks an isolation group to run in. if the workflow was started there, it'll attempt to pin it, unless there is an explicit
+// drain.
+func pickIsolationGroup(wfPartition defaultWorkflowPartitionConfig, available types.IsolationGroupConfiguration) string {
+	wfIG, isAvailable := available[wfPartition.WorkflowStartIsolationGroup]
+	if isAvailable && wfIG.State != types.IsolationGroupStateDrained {
+		return wfPartition.WorkflowStartIsolationGroup
+	}
+
+	// it's drained, fall back to picking a deterministic but random group
+	availableList := []string{}
+	for k, v := range available {
```

Comment:
`available` seems to only contain healthy isolation groups.

Reply:
That is the intention, yes? I'm not quite sure I understand your comment.?